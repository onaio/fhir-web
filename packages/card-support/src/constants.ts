// OPENSRP API endpoints
export const OPENSRP_URL_USER_ASSIGNMENT = '/organization/user-assignment';
export const OPENSRP_URL_CLIENT_SEARCH = '/client/search';
export const OPENSRP_URL_LOCATION_HIERARCHY = '/location/hierarchy';

/** CSV File MIME Types */
export const TEXT_CSV = 'text/csv';
export const APPLICATION_CSV = 'application/csv';
export const TEXT_PLAIN = 'text/plain';
export const TEXT_X_CSV = 'text/x-csv';
export const APPLICATION_VND_EXCEL = 'application/vnd.ms-excel';
export const APPLICATION_X_CSV = 'application/x-csv';
export const TEXT_COMMA_SEPARATED_VALUES = 'text/comma-separated-values';
export const TEXT_X_COMMA_SEPARATED_VALUES = 'text/x-comma-separated-values';
export const TEXT_TAB_SEPARATED_VALUES = 'text/tab-separated-values';
