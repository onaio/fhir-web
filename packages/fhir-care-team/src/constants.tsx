// url

// Query patams
export const SEARCH_QUERY_PARAM = 'searchQuery';

// FHIR API strings
export const URL_ADMIN = '/admin';
export const FHIR_CARE_TEAM = 'CareTeam';
export const FHIR_GROUPS = 'Group';
export const FHIR_PRACTITIONERS = 'Practitioner';
export const URL_EDIT_CARE_TEAM = `${URL_ADMIN}/CareTeam/edit`;
export const URL_CREATE_CARE_TEAM = `${URL_ADMIN}/CareTeam/new`;
export const URL_CARE_TEAM = `${URL_ADMIN}/CareTeams`;

// Route params
export const ROUTE_PARAM_CARE_TEAM_ID = 'careTeamId';
