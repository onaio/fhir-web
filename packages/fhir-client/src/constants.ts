export const dateFormat = 'yyyy-MM-dd';
