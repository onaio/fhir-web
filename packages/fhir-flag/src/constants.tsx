// urls
export const URL_ADMIN = '/admin';
export const URL_CLOSE_FLAGS = `${URL_ADMIN}/close-flags`;

// // magic strings

export const thatiMinutes = 30 * 60 * 1000;

// form fields
export const locationName = 'locationName' as const;
export const productName = 'productName' as const;
export const status = 'status' as const;
export const comments = 'comments' as const;
