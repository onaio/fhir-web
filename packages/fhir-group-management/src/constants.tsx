// url
export const ADD_EDIT_GROUP_URL = '/group/add-edit';
export const ADD_EDIT_COMMODITY_URL = '/commodity/add-edit';
export const LIST_GROUP_URL = '/groups/list';
export const LIST_COMMODITY_URL = '/commodity/list';

// magic strings
export const groupResourceType = 'Group';
export const listResourceType = 'List';

// form constants
export const id = 'id' as const;
export const identifier = 'identifier' as const;
export const name = 'name' as const;
export const active = 'active' as const;
export const type = 'type' as const;
export const unitOfMeasure = 'unitOfMeasure' as const;
