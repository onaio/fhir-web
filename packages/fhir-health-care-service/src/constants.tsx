// url
export const URL_ADD_HEALTHCARE = 'healthcare/add/';
export const URL_EDIT_HEALTHCARE = 'healthcare/edit/';

//
// Healthcares
//
export const HEALTH_CARE_SERVICE_RESOURCE_TYPE = 'HealthcareService';
export const HEALTHCARES_ENDPOINT = 'HealthcareService/';
