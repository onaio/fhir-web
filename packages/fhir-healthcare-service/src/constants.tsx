// url
export const ADD_EDIT_HEALTHCARE_SERVICE_URL = '/healthcare/add-edit';
export const LIST_HEALTHCARE_URL = '/healthcare/list';

//
// constants
//
export const healthCareServiceResourceType = 'HealthcareService';

// form constants
export const id = 'id' as const;
export const identifier = 'identifier' as const;
export const name = 'name' as const;
export const active = 'active' as const;
export const comment = 'comment' as const;
export const extraDetails = 'extraDetails' as const;
export const providedBy = 'providedBy' as const;
