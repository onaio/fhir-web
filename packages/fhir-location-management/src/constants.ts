// resource type  constants
export const locationHierarchyResourceType = 'LocationHierarchy';
export const locationResourceType = 'Location';

// Url
export const URL_ADMIN = '/admin';
export const URL_LOCATION_UNIT = `${URL_ADMIN}/location/unit`;
export const URL_LOCATION_UNIT_ADD = `${URL_LOCATION_UNIT}/add`;
export const URL_LOCATION_UNIT_EDIT = `${URL_LOCATION_UNIT}/edit`;
