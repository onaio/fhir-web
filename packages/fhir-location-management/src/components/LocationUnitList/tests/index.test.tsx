import React from 'react';
import { store } from '@opensrp/store';
import { LocationUnitList } from '..';
import { authenticateUser } from '@onaio/session-reducer';
import { QueryClient, QueryClientProvider } from 'react-query';
import nock from 'nock';
import {
  render,
  cleanup,
  waitForElementToBeRemoved,
  screen,
  within,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import { Router } from 'react-router';
import { createBrowserHistory } from 'history';
import { fhirHierarchy, onaOfficeSubLocation } from '../../../ducks/tests/fixtures';
import { Provider } from 'react-redux';
import { RoleContext } from '@opensrp/rbac';
import { superUserRole } from '@opensrp/react-utils';
import { locationHierarchyResourceType } from '../../../constants';
import { locationResourceType } from '../../../constants';
import userEvent from '@testing-library/user-event';
import { rootlocationFixture } from '../../RootLocationWizard/tests/fixtures';
import * as notifications from '@opensrp/notifications';

const history = createBrowserHistory();

jest.mock('fhirclient', () => {
  return jest.requireActual('fhirclient/lib/entry/browser');
});

const props = {
  fhirRootLocationId: 'eff94f33-c356-4634-8795-d52340706ba9',
  fhirBaseURL: 'http://test.server.org',
};

jest.mock('@opensrp/notifications', () => ({
  __esModule: true,
  ...Object.assign({}, jest.requireActual('@opensrp/notifications')),
}));

jest.setTimeout(10000);

nock.disableNetConnect();

describe('location-management/src/components/LocationUnitList', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        // react-query will retry failed request, this can cause some weird
        // behavior when testing a rejected request
        retry: false,
        cacheTime: 0,
      },
    },
  });

  const AppWrapper = (props) => {
    return (
      <Provider store={store}>
        <RoleContext.Provider value={superUserRole}>
          <Router history={history}>
            <QueryClientProvider client={queryClient}>
              <LocationUnitList {...props} />
            </QueryClientProvider>
          </Router>
        </RoleContext.Provider>
      </Provider>
    );
  };

  beforeAll(() => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        { api_token: 'hunter2', oAuth2Data: { access_token: 'hunter2', state: 'abcde' } }
      )
    );
  });

  afterEach(() => {
    nock.cleanAll();
    cleanup();
  });

  it('shows broken page', async () => {
    nock(props.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: props.fhirRootLocationId })
      .replyWithError({
        message: 'something awful happened',
        code: 'AWFUL_ERROR',
      });

    nock(props.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: 'missing' })
      .reply(200, {});

    const { rerender } = render(<AppWrapper {...props} />);

    await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

    expect(screen.getByText(/reason: something awful happened/)).toBeInTheDocument();

    rerender(<AppWrapper {...{ ...props, fhirRootLocationId: 'missing' }} />);

    await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

    expect(screen.getByText(/404/)).toBeInTheDocument();
    expect(
      screen.getByText(/Sorry, the resource you requested for, does not exist/)
    ).toBeInTheDocument();
  });

  it('works correctly', async () => {
    nock(props.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: props.fhirRootLocationId })
      .reply(200, fhirHierarchy)
      .persist();

    nock(props.fhirBaseURL).get('/Location/303').reply(200, onaOfficeSubLocation);

    render(<AppWrapper {...props} />);

    await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

    expect(screen.getByText(/Location Unit Management/)).toBeInTheDocument();

    // initially show single user defined root location
    const treeSection = document.querySelector('.ant-tree') as HTMLElement;
    const firstRootLoc = within(treeSection).getByTitle(/Ona Office Sub Location/);
    expect(firstRootLoc).toMatchSnapshot('user defined root location');

    // table row should be one containing information
    const table = document.querySelector('table');
    // how many body rows - should be 1
    expect(table.querySelectorAll('tbody tr')).toHaveLength(1);
    const firstRowTd = table.querySelectorAll('tbody tr:nth-child(1) td');
    firstRowTd.forEach((td) => {
      expect(td).toMatchSnapshot('first row containing ona office loc details');
    });

    // view details
    const hamburger = document.querySelector('.more-options');
    fireEvent.click(hamburger);
    // should see the popup with view details
    expect(document.querySelector('[data-testid="view-location"]')).toBeInTheDocument();
    fireEvent.click(document.querySelector('[data-testid="view-location"]'));

    // should load single location
    await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

    // see location details in view details
    const viewDetailsSection = document.querySelector('[data-testid="view-details"]');
    viewDetailsSection.querySelectorAll('[data-testid="single-key-value"]').forEach((keyValue) => {
      expect(keyValue).toMatchSnapshot('key value');
    });

    // close view details section
    fireEvent.click(viewDetailsSection.querySelector('button.float-right'));
    expect(document.querySelector('[data-testid="view-details"]')).not.toBeInTheDocument();

    // lets expand on tree search and select a node
    const antTree = document.querySelector('.ant-tree-list');
    const firstDownCaret = antTree.querySelector('.anticon-caret-down');
    fireEvent.click(firstDownCaret);

    // tree title visible now - represents nested nodes that are now visible
    document.querySelectorAll('.ant-tree-title').forEach((nodeTitle) => {
      expect(nodeTitle).toMatchSnapshot('visible nodes on expand');
    });

    const lastTextNode = screen.getByText('Part Of Sub Location');
    fireEvent.click(lastTextNode);

    // tree title visible now - represents nested nodes that are now visible
    document.querySelectorAll('.ant-tree-title').forEach((nodeTitle) => {
      expect(nodeTitle).toMatchSnapshot('visible nodes on expanding part of sub location');
    });

    // table change
    expect(document.querySelectorAll('table tbody tr')).toHaveLength(5);

    // deselect node
    fireEvent.click(screen.getByTitle('Part Of Sub Location'));

    // table change- node deselect
    expect(document.querySelectorAll('table tbody tr')).toHaveLength(1);

    // invalidate queries to initiate a refetch of locationhierarchy
    queryClient.invalidateQueries([locationHierarchyResourceType]).catch((err) => {
      throw err;
    });

    await waitFor(() => {
      expect(screen.getByText(/Refreshing data/i)).toBeInTheDocument();
    });
    await waitForElementToBeRemoved(screen.getByText(/Refreshing data/i));
  });

  it('Passes selected node as the parent location when adding location clicked', async () => {
    nock(props.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: props.fhirRootLocationId })
      .reply(200, fhirHierarchy)
      .persist();

    render(<AppWrapper {...props} />);

    await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

    expect(screen.getByText(/Location Unit Management/)).toBeInTheDocument();

    // initially show single user defined root location
    const treeSection = document.querySelector('.ant-tree') as HTMLElement;
    const firstRootLoc = within(treeSection).getByTitle(/Ona Office Sub Location/);
    expect(firstRootLoc).toMatchSnapshot('user defined root location');

    // click/select this first root location
    userEvent.click(firstRootLoc);

    // then click add location
    const addLocationBtn = screen.getByText(/Add Location Unit/);
    userEvent.click(addLocationBtn);

    // check where we redirected to
    expect(history.location.search).toEqual('?parentId=Location%2F303');
  });

  it('Root location wizard works correclty', async () => {
    const notificationSuccessMock = jest.spyOn(notifications, 'sendSuccessNotification');

    nock(props.fhirBaseURL)
      .get(`/${locationHierarchyResourceType}/_search`)
      .query({ _id: props.fhirRootLocationId })
      .reply(404, {
        resourceType: 'OperationOutcome',
        issue: [
          {
            severity: 'error',
            code: 'processing',
            diagnostics: `HAPI-2001: Resource Location/${props.fhirRootLocationId} is not known`,
          },
        ],
      });

    nock(props.fhirBaseURL)
      .get(`/${locationResourceType}/_search`)
      .query({ _summary: 'count' })
      .reply(200, { total: 0 })
      .persist();

    nock(props.fhirBaseURL)
      .put(`/Location/${rootlocationFixture.id}`, rootlocationFixture)
      .reply(201, {})
      .persist();

    render(<AppWrapper {...props} />);

    await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

    expect(screen.getByText(/Location Unit Management/)).toBeInTheDocument();
    expect(screen.getByText(/Root location was not found/)).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByText(/No locations have been uploaded yet./)).toBeInTheDocument();
    });

    const locationCreate = screen.getByRole('button', { name: 'Create root location.' });

    fireEvent.click(locationCreate);

    expect(screen.getByText(/This action will create a new location with id/)).toBeInTheDocument();
    const confirmBtn = screen.getByRole('button', { name: 'Proceed' });

    fireEvent.click(confirmBtn);

    await waitFor(() => {
      expect(notificationSuccessMock).toHaveBeenCalledWith('Root location uploaded to the server.');
    });
  });
});
