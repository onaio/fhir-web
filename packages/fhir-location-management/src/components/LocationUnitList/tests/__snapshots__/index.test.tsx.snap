// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`location-management/src/components/LocationUnitList works correctly: first row containing ona office loc details 1`] = `
<td
  class="ant-table-cell"
>
  Ona Office Sub Location
</td>
`;

exports[`location-management/src/components/LocationUnitList works correctly: first row containing ona office loc details 2`] = `
<td
  class="ant-table-cell"
>
  Root FHIR Location
</td>
`;

exports[`location-management/src/components/LocationUnitList works correctly: first row containing ona office loc details 3`] = `
<td
  class="ant-table-cell"
>
  Jurisdiction
</td>
`;

exports[`location-management/src/components/LocationUnitList works correctly: first row containing ona office loc details 4`] = `
<td
  class="ant-table-cell"
>
  active
</td>
`;

exports[`location-management/src/components/LocationUnitList works correctly: first row containing ona office loc details 5`] = `
<td
  class="ant-table-cell"
>
  <span
    class="d-flex justify-content-end align-items-center Actions"
  >
    <a
      href="/admin/location/unit/edit/303"
    >
      <button
        class="ant-btn ant-btn-link m-0 p-1"
        type="button"
      >
        <span>
          Edit
        </span>
      </button>
    </a>
    <div
      class="ant-divider ant-divider-vertical"
      role="separator"
    />
    <span
      aria-label="more"
      class="anticon anticon-more ant-dropdown-trigger more-options"
      role="img"
      tabindex="-1"
    >
      <svg
        aria-hidden="true"
        data-icon="more"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
        />
      </svg>
    </span>
  </span>
</td>
`;

exports[`location-management/src/components/LocationUnitList works correctly: key value 1`] = `
<div
  class="pb-3"
  data-testid="single-key-value"
>
  <dl
    class="singleKeyValue"
  >
    <div
      class="singleKeyValue-pair"
    >
      <dt
        class="singleKeyValue-pair__label"
      >
        <span
          class="ant-typography ant-typography-secondary"
        >
          Name
        </span>
      </dt>
      <dd
        class="singleKeyValue-pair__value"
      >
        <span
          class="ant-typography"
        >
          Ona Office Sub Location
        </span>
      </dd>
    </div>
  </dl>
</div>
`;

exports[`location-management/src/components/LocationUnitList works correctly: key value 2`] = `
<div
  class="pb-3"
  data-testid="single-key-value"
>
  <dl
    class="singleKeyValue"
  >
    <div
      class="singleKeyValue-pair"
    >
      <dt
        class="singleKeyValue-pair__label"
      >
        <span
          class="ant-typography ant-typography-secondary"
        >
          Status
        </span>
      </dt>
      <dd
        class="singleKeyValue-pair__value"
      >
        <span
          class="ant-typography"
        >
          active
        </span>
      </dd>
    </div>
  </dl>
</div>
`;

exports[`location-management/src/components/LocationUnitList works correctly: key value 3`] = `
<div
  class="pb-3"
  data-testid="single-key-value"
>
  <dl
    class="singleKeyValue"
  >
    <div
      class="singleKeyValue-pair"
    >
      <dt
        class="singleKeyValue-pair__label"
      >
        <span
          class="ant-typography ant-typography-secondary"
        >
          Description
        </span>
      </dt>
      <dd
        class="singleKeyValue-pair__value"
      >
        <span
          class="ant-typography"
        >
          The Sub location
        </span>
      </dd>
    </div>
  </dl>
</div>
`;

exports[`location-management/src/components/LocationUnitList works correctly: key value 4`] = `
<div
  class="pb-3"
  data-testid="single-key-value"
>
  <dl
    class="singleKeyValue"
  >
    <div
      class="singleKeyValue-pair"
    >
      <dt
        class="singleKeyValue-pair__label"
      >
        <span
          class="ant-typography ant-typography-secondary"
        >
          Physical Type
        </span>
      </dt>
      <dd
        class="singleKeyValue-pair__value"
      >
        <span
          class="ant-typography"
        >
          Jurisdiction
        </span>
      </dd>
    </div>
  </dl>
</div>
`;

exports[`location-management/src/components/LocationUnitList works correctly: key value 5`] = `
<div
  class="pb-3"
  data-testid="single-key-value"
>
  <dl
    class="singleKeyValue"
  >
    <div
      class="singleKeyValue-pair"
    >
      <dt
        class="singleKeyValue-pair__label"
      >
        <span
          class="ant-typography ant-typography-secondary"
        >
          Alias
        </span>
      </dt>
      <dd
        class="singleKeyValue-pair__value"
      >
        <span
          class="ant-typography"
        >
          ona office
        </span>
      </dd>
    </div>
  </dl>
</div>
`;

exports[`location-management/src/components/LocationUnitList works correctly: user defined root location 1`] = `
<span
  title="Ona Office Sub Location"
>
  Ona Office Sub Location
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expand 1`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Ona Office Sub Location"
  >
    Ona Office Sub Location
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expand 2`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Level 0"
  >
    Level 0
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expand 3`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Arundel mobile clinic"
  >
    Arundel mobile clinic
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expand 4`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Part Of Sub Location"
  >
    Part Of Sub Location
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 1`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Ona Office Sub Location"
  >
    Ona Office Sub Location
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 2`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Level 0"
  >
    Level 0
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 3`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Arundel mobile clinic"
  >
    Arundel mobile clinic
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 4`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Part Of Sub Location"
  >
    Part Of Sub Location
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 5`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Nairobii"
  >
    Nairobii
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 6`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Test Sub Loc"
  >
    Test Sub Loc
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 7`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Test Loc 113"
  >
    Test Loc 113
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 8`] = `
<span
  class="ant-tree-title"
>
  <span
    title="Test Loc 1145"
  >
    Test Loc 1145
  </span>
</span>
`;

exports[`location-management/src/components/LocationUnitList works correctly: visible nodes on expanding part of sub location 9`] = `
<span
  class="ant-tree-title"
>
  <span
    title="MOH HQ 2"
  >
    MOH HQ 2
  </span>
</span>
`;
